pandaserver/dataservice/ddm.py

```diff
@@ -0,0 +1,865 @@
+                vuid = f"{vuid[0:8]}-{vuid[8:12]}-{vuid[12:16]}-{vuid[16:20]}-{vuid[20:32]}"
+                duid = vuid
+                # add /
+                if dataset_name.endswith("/") and not name.endswith("/"):
+                    name += "/"
+                if old or ":" not in dataset_name:
+                    key_name = name
+                else:
+                    key_name = str(f"{scope}:{name}")
+                if key_name not in result:
+                    result[key_name] = {"duid": duid, "vuids": [vuid]}
+            return result, ""
+        except Exception as error:
+            return None, f"{str(error)} {traceback.format_exc()}"
+
+    # list datasets in container
+    def list_datasets_in_container(self, container_name: str):
+        """
+        List datasets in a Rucio container.
+
+        This method lists datasets in a Rucio container by extracting the scope from the container name and getting the DIDs (Data Identifiers).
+        It generates a unique identifier (vuid) for each DID and stores it in a list.
+        If an exception occurs during the process, it returns None and the exception message.
+
+        Parameters:
+        container_name (str): The name of the container.
+
+        Returns:
+        Tuple[Union[None, List[str]], str]: A tuple containing a list of datasets and a string message.
+        If an exception occurs, the list is None and the string contains the exception message.
+        """
+        result = []
+        # extract scope from dataset
+        scope, container_name = self.extract_scope(container_name)
+        if container_name.endswith("/"):
+            container_name = container_name[:-1]
+        try:
+            # get dids
+            client = RucioClient()
+            for i in client.list_content(scope, container_name):
+                if i["type"] == "DATASET":
+                    result.append(str(f"{i['scope']}:{i['name']}"))
+            return result, ""
+        except Exception:
+            err_type, err_value = sys.exc_info()[:2]
+            return None, f"{err_type} {err_value}"
+
+    # list dataset replicas
+    def list_dataset_replicas(self, dataset_name: str):
+        """
+        List dataset replicas in Rucio.
+
+        This method lists dataset replicas in Rucio by extracting the scope from the dataset name and getting the replicas.
+        It stores the replicas in a dictionary and returns it along with a status code.
+        If an exception occurs during the process, it returns the error type and value.
+
+        Parameters:
+        dataset_name (str): The name of the dataset.
+
+        Returns:
+        Tuple[int, Union[str, dict]]: A tuple containing a status code and a dictionary of dataset replicas or an error message.
+        If an exception occurs, the dictionary is None and the string contains the error message.
+        """
+        ret_map = {}
```

Comment:
ret_map is the universal return variable in panda, but it's not a very clear name